sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1583 @@
+            self._on_X_parallel_init(thread_num)
+
+            for X_chunk_idx in prange(self.X_n_chunks, schedule='static'):
+                X_start = X_chunk_idx * self.X_n_samples_chunk
+                if (X_chunk_idx == self.X_n_chunks - 1
+                    and self.X_n_samples_remainder > 0):
+                    X_end = X_start + self.X_n_samples_remainder
+                else:
+                    X_end = X_start + self.X_n_samples_chunk
+
+                # Reinitializing thread datastructures for the new X chunk
+                self._on_X_prange_iter_init(thread_num, X_start, X_end)
+
+                for Y_chunk_idx in range(self.Y_n_chunks):
+                    Y_start = Y_chunk_idx * self.Y_n_samples_chunk
+                    if (Y_chunk_idx == self.Y_n_chunks - 1
+                        and self.Y_n_samples_remainder > 0):
+                        Y_end = Y_start + self.Y_n_samples_remainder
+                    else:
+                        Y_end = Y_start + self.Y_n_samples_chunk
+
+                    self._reduce_on_chunks(
+                        X_start, X_end,
+                        Y_start, Y_end,
+                        thread_num,
+                    )
+
+                # Adjusting thread datastructures on the full pass on Y
+                self._on_X_prange_iter_finalize(thread_num, X_start, X_end)
+
+            # end: for X_chunk_idx
+
+            # Deallocating thread datastructures
+            self._on_X_parallel_finalize(thread_num)
+
+        # end: with nogil, parallel
+        return
+
+    @final
+    cdef void _parallel_on_Y(self) nogil:
+        """Computes the reduction of each vector (row) of X on Y
+        by parallelizing computation on chunks of Y.
+
+        This strategy dispatches chunks of Y uniformly on threads.
+        Each thread then iterates on all the chunks of X. This strategy is
+        embarrassingly parallel but uses intermediate datastructures
+        synchronisation. However it is more useful in practice (because Y is
+        larger than X generally).
+
+        Private datastructures are modified internally by threads.
+
+        Private template methods can be implemented on subclasses to
+        interact with those datastructures at various stages.
+        """
+        cdef:
+            ITYPE_t Y_start, Y_end, X_start, X_end, X_chunk_idx, Y_chunk_idx
+            ITYPE_t num_threads = min(self.Y_n_chunks, self.effective_omp_n_thread)
+            ITYPE_t thread_num
+
+        # Allocating datastructures
+        self._on_Y_init(num_threads)
+
+        for X_chunk_idx in range(self.X_n_chunks):
+            X_start = X_chunk_idx * self.X_n_samples_chunk
+            if X_chunk_idx == self.X_n_chunks - 1 and self.X_n_samples_remainder > 0:
+                X_end = X_start + self.X_n_samples_remainder
+            else:
+                X_end = X_start + self.X_n_samples_chunk
+
+            with nogil, parallel(num_threads=num_threads):
```

Comment:
Currently, `_parallel_on_Y` uses 3 OpenMP parallel regions (here and in `_on_Y_after_parallel` and `_on_Y_finalize` overridings) with various number of threads (`num_threads` here, `self.effective_omp_n_thread` in other cases.

A single parallel region is preferable (because setting one comes with overhead). In that case and to make use of all threads, we probably want to use `self.effective_omp_n_thread` even if `self.Y_n_chunks` is smaller.

Reply:
I am not sure that this introduces a lot of overhead but if the code stays easy to follow, why not. I am not sure about the impact on the memory usage w.r.t. the thread-local buffers when the number of threads are large, though. It's worth trying and bench marking.

Reply:
It looks like merging sections can help because a [`flush`](https://www.openmp.org/spec-html/5.0/openmpsu96.html) happens after each of them¹ and this just kills memory locality.

[1]: https://www.openmp.org/spec-html/5.0/openmpsu96.html

Reply:
As discussed, it might be possible to rework this in another PR. Let's keep this simple for this one: performances are already satisfying.